internal/validate/validate.go

```diff
@@ -0,0 +1,134 @@
+// Package validate internalizes Go Playground's Validator framework, setting
+// some common options that we use everywhere, providing some useful helpers,
+// and exporting a simplified API.
+package validate
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/go-playground/validator/v10"
+)
+
+// WithRequiredStructEnabled can be removed once validator/v11 is released.
+var validate = validator.New(validator.WithRequiredStructEnabled()) //nolint:gochecknoglobals
+
+func init() { //nolint:gochecknoinits
+	validate.RegisterTagNameFunc(preferPublicName)
+}
+
+// PublicFacingMessage builds a complete error message from a validator error
+// that's suitable for public-facing consumption.
+//
+// I only added a few possible validations to start. We'll probably need to add
+// more as we go and expand our usage.
+func PublicFacingMessage(validatorErr error) string {
+	var message string
+
+	//nolint:errorlint
+	if validationErrs, ok := validatorErr.(validator.ValidationErrors); ok {
+		for _, fieldErr := range validationErrs {
+			switch fieldErr.Tag() {
+			case "lte":
+				fallthrough // lte and max are synonyms
+			case "max":
+				kind := fieldErr.Kind()
+				if kind == reflect.Ptr {
+					kind = fieldErr.Type().Elem().Kind()
+				}
+
+				switch kind {
+				case reflect.Float32:
+					fallthrough
+				case reflect.Float64:
+					fallthrough
+				case reflect.Int:
+					fallthrough
+				case reflect.Int32:
+					fallthrough
+				case reflect.Int64:
+					message += fmt.Sprintf(" Field `%s` must be less than or equal to %s.",
+						fieldErr.Field(), fieldErr.Param())
```

Comment:
do you prefer this style with `fallthrough` over putting all the values in a single `case` clause?

Reply:
I find it nominally easier to read because it's less cramped, but no strong opinion either way. The one line with commas is probably more conventional, so I changed it to that.